drivers/QSPI.cpp

```diff
@@ -0,0 +1,271 @@
+#define IS_SIZE_VALID(size)                ((size == QSPI_CFG_ADDR_SIZE_NONE) || (size == QSPI_CFG_ADDR_SIZE_8) || (size == QSPI_CFG_ADDR_SIZE_16) || (size == QSPI_CFG_ADDR_SIZE_24) || (size == QSPI_CFG_ADDR_SIZE_32))      
+#define IS_ALT_SIZE_VALID(alt_size)        ((alt_size == QSPI_CFG_ALT_SIZE_NONE) || (alt_size == QSPI_CFG_ALT_SIZE_8) || (alt_size == QSPI_CFG_ALT_SIZE_16) || (alt_size == QSPI_CFG_ALT_SIZE_24) || (alt_size == QSPI_CFG_ALT_SIZE_32))      
+
+namespace mbed {
+
+QSPI* QSPI::_owner = NULL;
+SingletonPtr<PlatformMutex> QSPI::_mutex; 
+    
+QSPI::QSPI(PinName io0, PinName io1, PinName io2, PinName io3, PinName sclk, PinName ssel) :
+        _qspi() {
+    _qspi_io0 = io0;
+    _qspi_io1 = io1;
+    _qspi_io2 = io2;
+    _qspi_io3 = io3;
+    _qspi_clk = sclk;
+    _qspi_cs = ssel;
+    _inst_width = QSPI_CFG_BUS_SINGLE;
+    _address_width = QSPI_CFG_BUS_SINGLE;
+    _address_size = QSPI_CFG_ADDR_SIZE_24;
+    _alt_width = QSPI_CFG_BUS_SINGLE;
+    _alt_size = QSPI_CFG_ALT_SIZE_NONE;
+    _data_width = QSPI_CFG_BUS_SINGLE;
+    _num_dummy_cycles = 0;
+    _mode = 0;
+    _hz = ONE_MHZ;             
+}
+        
+qspi_status_t QSPI::configure_format(qspi_bus_width_t inst_width, qspi_bus_width_t address_width, qspi_address_size_t address_size, qspi_bus_width_t alt_width, qspi_alt_size_t alt_size, qspi_bus_width_t data_width, int dummy_cycles, int mode ) {
+    if(!IS_BUS_WIDTH_VALID(inst_width)) 
+        return QSPI_STATUS_INVALID_PARAMETER;
+    if(!IS_BUS_WIDTH_VALID(address_width)) 
+        return QSPI_STATUS_INVALID_PARAMETER;
+    if(!IS_SIZE_VALID(address_size)) 
+        return QSPI_STATUS_INVALID_PARAMETER;
+    if(!IS_BUS_WIDTH_VALID(alt_width)) 
+        return QSPI_STATUS_INVALID_PARAMETER;
+    if(!IS_ALT_SIZE_VALID(alt_size)) 
+        return QSPI_STATUS_INVALID_PARAMETER;
+    if(!IS_BUS_WIDTH_VALID(data_width)) 
+        return QSPI_STATUS_INVALID_PARAMETER;
+    if(dummy_cycles < 0) 
+        return QSPI_STATUS_INVALID_PARAMETER;                         
+    if(mode != 0 && mode != 1) 
+        return QSPI_STATUS_INVALID_PARAMETER;                            
+    
+    lock();
+    _inst_width = inst_width;
+    _address_width = address_width;
+    _address_size = address_size;
+    _alt_width = alt_width;
+    _alt_size = alt_size;
+    _data_width = data_width; 
+    _num_dummy_cycles = dummy_cycles;
+    _mode = mode;
+    unlock();
+    
+    return QSPI_STATUS_OK;
+}
+
+qspi_status_t QSPI::set_frequency(int hz) {
+    qspi_status_t ret_status = QSPI_STATUS_OK; 
+    lock();
+    _hz = hz;
+    
+    //If the same owner, just change freq.
+    //Otherwise we may have to change mode as well, so call _acquire
+    if (_owner == this) {
+        if(QSPI_STATUS_OK != qspi_frequency(&_qspi, _hz)) {
+            ret_status = QSPI_STATUS_ERROR;
+        }
+    } else {
+        _acquire();
+    }
+    unlock();
+    
+    return ret_status;
+}
+
+qspi_status_t QSPI::initialize() {
```

Comment:
I still believe init() should be done in ctor, to follow how we do it in driver

Reply:
I think we will call qspi_init in configure_format itself and an internal variable to track if its properly init-ed.